zero/gyrokinetic_maxwellian_moments.c

```diff
@@ -0,0 +1,233 @@
+  up->num_conf_basis = inp->conf_basis->num_basis;
+  // Determine factor to divide out of temperature computation
+  // If 1x1v, up->vdim = 1, otherwise up->vdim = 3
+  if (up->phase_basis.ndim - up->conf_basis.ndim == 1) {
+    up->vdim = up->phase_basis.ndim - up->conf_basis.ndim;
+  }
+  else {
+    up->vdim = 3;
+  }
+  up->gk_geom = gkyl_gk_geometry_acquire(inp->gk_geom);
+  up->divide_jacobgeo = inp->divide_jacobgeo;
+
+  long conf_range_ncells = inp->conf_range->volume;
+  long conf_range_ext_ncells = inp->conf_range_ext->volume;
+
+  if (inp->use_gpu) {
+    up->M0 = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->M1 = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->u_par = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->u_par_dot_M1 = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->pressure = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->temperature = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    // Bin op memory for needed weak divisions
+    up->mem = gkyl_dg_bin_op_mem_cu_dev_new(conf_range_ncells, up->num_conf_basis);
+    if (up->vdim == 3) {
+      // Additional moments if computing Bi-Maxwellian moments (Tpar, Tperp)
+      up->p_par = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+      up->t_par = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);    
+      up->p_perp = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+      up->t_perp = gkyl_array_cu_dev_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells); 
+    }
+  }
+  else {
+    up->M0 = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->M1 = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->u_par = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->u_par_dot_M1 = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->pressure = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    up->temperature = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    // Bin op memory for needed weak divisions
+    up->mem = gkyl_dg_bin_op_mem_new(conf_range_ncells, up->num_conf_basis);
+    if (up->vdim == 3) {
+      // Additional moments if computing Bi-Maxwellian moments (Tpar, Tperp)
+      up->p_par = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+      up->t_par = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+      up->p_perp = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+      up->t_perp = gkyl_array_new(GKYL_DOUBLE, up->num_conf_basis, conf_range_ext_ncells);
+    }
+  }
```

Comment:
@JunoRavin  L35-L68: can you create a local mkarr function like the one in the apps, and consolidate these two blocks into a single one please?